### DISCUSSION TITLE
Feature idea: user-definable buttons in the Playtime mixer that can be assigned actions

### DISCUSSION THREAD

**Discussion by TubularCorpration** [helgoboss/helgobox#1107] [Ideas]
Nothing too complex (in theory, might be hard to implement) - in the empty space to the right of the fader in each column (opposite the arm, mute and solo buttons), add a few buttons that can have any available action assigned to them. Pressing the button will select the corresponding Reaper track (since a lot of useful actions apply to the current track - in general I think the option to make selection a column in Playtime's TCP also select the corresponding track in the Reaper TCP the way it does when you select a column in the clip matrix would be handy) and then run the assigned action.

It might make sense to also have a a few user definable buttons somewhere else on the UI (maybe one or two on each side of the Playtime logo, since there's empty space already) for actions that are independent of track selection, just to keep things more logically organized.

Graphically they could be as simple as one letter or number (a user-definable tooltip might be a good idea here) and functionally they could work exactly the same way the buttons on Reaper's toolbars.

This would let you access a lot of deeper Reaper functions, scripts and extensions but keep Playtime in full screen so the workflow wouldn't be interrupted.  There's so much that can be done with user scripts and custom actions that it would be a pretty simple way to add a quite a bit of customizability to Playtime without making any significant changes to the UI or core workflow.

The specific use case that made me think of this is that I would love to be able to run the fantastic [Paranormal FX Router](https://forum.cockos.com/showthread.php?t=283054) script directly from the Playtime mixer instead of using its FX button to open the standard Reaper plugin browser, but there are tons of other things I can think of that would be really handy to have available as buttons in a full-screen Playtime session.

A few more examples off the top of my head:

- Run [Quick Adder 2](https://forum.cockos.com/showthread.php?t=232928) to load effects chains, access the full action list, etc. while Playtime is in fullscreen.
- Mute/unmute a track send from a touch screen for live dub echo
- Run one of the many plugin randomizer scripts from a touchscreen as a live performance tool
- Run a custom script to increment (or decrement) and send a MIDI program change message to the selected track, to scroll through presets on a hardware synth or VSTi from within Playtime (could also be done with a custom action made up of some SWS actions and ReaControlMIDI)

**Comment by helgoboss** [comment]
Thanks. Interesting proposal and already very well thought-out. Shouldn't be too difficult to add. One question comes up: This should probably be a global thing influencing all Playtime instances?

**Comment by TubularCorpration** [comment]
That's a good question, I think it would make more sense for it to be global, although I can also envision situations where it would make sense for it to be per-instance so you could customize an instance for a specific project without .  

It would be a little more work, but maybe some kind of system of button sets (which could be a simple JSON file with a list of all the assignable buttons, their associated command IDs, and maybe a flag for momentary or toggle).  At first I was thinking of full button sets, but that might get tricky to scale to different numbers/types of columns.

I think the way to go would be user-definable defaults for global buttons, audio column buttons and MIDI column buttons (to match the two column types available in the templates that are available when you start Playtime) and maybe a simple preset system that would let users save and load different sets of button assignments for a single column.  The defaults would load with every new Reaper instance and new columns would start with the default button assignments (maybe a user setting for whether new columns defaulted to the audio or MIDI button presets).

So by default every instance of Playtime would start with the same default user buttons for consistency across instances, but users would also be able to override the defaults ON A PER-INSTANCE basis by reassigning buttons manually or loading different user presets per column.

There would need to be some kind of system to save button assignments so that they would persist when Reaper was closed and restarted anyway and it seems to me like that would be the bulk of the work to implement this whole idea, so why not expose that as a user-button preset system.  Everything (button assignment, toggle/momentary selection, and preset save/load) could be added to the Global and Column playback settings, so other than the actual buttons it wouldn't affect the UI much at all.

The more I think this through, the more it seems like it would add a lot of power while not being too complicated to implement - there are so many amazing third party scripts and extensions that would suddenly become part of the Playtime workflow.  In the longer term, if you ever added a docker to the Playtime UI (I think you mentioned that in one of the threads on the cockos forums at some point) having user buttons that opened docked extensions would make it possible to approximate (and go way beyond) the Ableton Live UI (i.e. by opening [LBX Stripper](https://forums.cockos.com/showthread.php?p=2581389) docked to Playtime as a stand-in for Ableton's "rack" system) without having to actually implement anything but the docker and user buttons.  EDIT: if I'm actually thinking of your plans to make Playtime dockable in Reaper as opposed to adding a docker to the Playtime UI, you could add an option for a user button to open an assigned script or extension in the Reaper docker and achieve pretty much the same thing, but this is all stuff to think about later if the button system I originally described is already implemented and working well)
